dag_in_context/src/schedule.rs

```diff
@@ -2,41 +2,47 @@ pub(crate) fn helpers() -> String {
     "
     ;; first, run substitution and drop to saturation
     ;; these depend on type analysis, always-run, and context
+
+    ;; first, saturate always run
     (saturate
-        (saturate
-          type-analysis
-          (saturate type-helpers)
-          error-checking
-          always-run
-          context)
-        
-        (saturate drop)
-        apply-drop-unions
-        cleanup-drop
+        (saturate 
+            (saturate type-helpers)
+            type-analysis)
+        (saturate 
+            (saturate type-helpers)
+            always-run)
+        error-checking)
+
+    ;; second, saturate substitution and drop
+    ;; which depend on context and is-resolved
+    (saturate
+        (saturate 
+            (saturate type-helpers)
+            type-analysis)
+        (saturate is-resolved)
 
         (saturate subst)
         apply-subst-unions
-        cleanup-subst)
+        cleanup-subst
+        (saturate context)
 
-    ;; now run canonicalization helpers, interval analysis
-    (saturate
-        (saturate
-          type-analysis
-          (saturate type-helpers))
-        error-checking
-        always-run
-        canon
-        interval-analysis
-        always-switch-rewrite
-        loop-iters-analysis
-        ; memory-helpers TODO run memory helpers for memory optimizations
+        (saturate drop)
+        apply-drop-unions
+        cleanup-drop
     )
 
+    (saturate canon)
+    (saturate interval-analysis)
```

Comment:
do we need to do types again here? Idk if interval analysis depends on canon

Reply:
`canon` is just doing some equational rewriting, so I think this just means we are missing some type and interval information for the new subexpressions generated in the current round, which probably won't matter I think